homeassistant/components/sense/sensor.py

```diff
@@ -46,64 +54,172 @@ def __init__(self, name, sensor_type):
-                update_call = update_active
-            else:
-                update_call = update_trends
 
             unique_id = f"{sense_monitor_id}-{type_id}-{var}".lower()
             devices.append(
-                Sense(
-                    data, name, sensor_type, is_production, update_call, var, unique_id
+                SenseTrendsSensor(
+                    data,
+                    name,
+                    sensor_type,
+                    is_production,
+                    update_trends,
+                    var,
+                    unique_id,
                 )
             )
 
     async_add_entities(devices)
 
 
-class Sense(Entity):
+class SenseActiveSensor(Entity):
     """Implementation of a Sense energy sensor."""
 
     def __init__(
-        self, data, name, sensor_type, is_production, update_call, sensor_id, unique_id
+        self,
+        data,
+        name,
+        sensor_type,
+        is_production,
+        sense_monitor_id,
+        sensor_id,
+        unique_id,
     ):
         """Initialize the Sense sensor."""
         name_type = PRODUCTION_NAME if is_production else CONSUMPTION_NAME
         self._name = f"{name} {name_type}"
         self._unique_id = unique_id
         self._available = False
         self._data = data
+        self._sense_monitor_id = sense_monitor_id
         self._sensor_type = sensor_type
-        self.update_sensor = update_call
         self._is_production = is_production
         self._state = None
+        self._undo_dispatch_subscription = None
 
-        if sensor_type == ACTIVE_TYPE:
-            self._unit_of_measurement = POWER_WATT
+    @property
+    def name(self):
+        """Return the name of the sensor."""
+        return self._name
+
+    @property
+    def state(self):
+        """Return the state of the sensor."""
+        return self._state
+
+    @property
+    def available(self):
+        """Return the availability of the sensor."""
+        return self._available
+
+    @property
+    def unit_of_measurement(self):
+        """Return the unit of measurement of this entity, if any."""
+        return POWER_WATT
+
+    @property
+    def icon(self):
+        """Icon to use in the frontend, if any."""
+        return ICON
+
+    @property
+    def unique_id(self):
+        """Return the unique id."""
+        return self._unique_id
+
+    @property
+    def should_poll(self):
+        """Return the device should not poll for updates."""
+        return False
+
+    async def async_update(self):
+        """Get the latest data, update state."""
+        if self._is_production:
+            self._state = round(self._data.active_solar_power)
         else:
-            self._unit_of_measurement = ENERGY_KILO_WATT_HOUR
+            self._state = round(self._data.active_power)
+
+        self._available = True
+
+    async def async_added_to_hass(self):
+        """Register callbacks."""
+
+        @callback
+        def update():
```

Comment:
So what would be faster here is to:
 - have this method live on the instance, not be defined inline
 - have this method update the attributes that is currently done in `async_update`
 - call `self.async_write_ha_state()`

Now you are saving a new task from being written and is the state directly written.

Reply:
Thanks. This is a much better way.  Should reduce the calls by a lot.